xwiki-rendering-transformations/xwiki-rendering-transformation-macro/src/main/java/org/xwiki/rendering/internal/transformation/macro/HTMLRawBlockFilter.java

```diff
@@ -0,0 +1,171 @@
+     */
+    private String cleanHTML(String content, MacroTransformationContext context, boolean restricted,
+        Syntax targetSyntax)
+        throws MacroExecutionException
+    {
+        String cleanedContent = content;
+
+        HTMLCleanerConfiguration cleanerConfiguration = getCleanerConfiguration(restricted, targetSyntax);
+
+        // Note that we trim the content since we want to be lenient with the user in case he has entered
+        // some spaces/newlines before a XML declaration (prolog). Otherwise the XML parser would fail to parse.
+        Document document = this.htmlCleaner.clean(new StringReader(cleanedContent), cleanerConfiguration);
+
+        // Since XML can only have a single root node and since we want to allow users to put
+        // content such as the following, we need to wrap the content in a root node:
+        // <tag1>
+        // ..
+        // </tag1>
+        // <tag2>
+        // </tag2>
+        // In addition we also need to ensure the XHTML DTD is defined so that valid XHTML entities can be
+        // specified.
+
+        // Remove the HTML envelope since this macro is only a fragment of a page which will already have an
+        // HTML envelope when rendered. We remove it so that the HTML <head> tag isn't output.
+        HTMLUtils.stripHTMLEnvelope(document);
+
+        // If in inline mode verify we have inline HTML content and remove the top level paragraph if there's one
+        if (context.isInline()) {
+            // TODO: Improve this since when're inside a table cell or a list item we can allow non inline items too
+            Element root = document.getDocumentElement();
+            if (root.getChildNodes().getLength() == 1 && root.getFirstChild().getNodeType() == Node.ELEMENT_NODE
+                && root.getFirstChild().getNodeName().equalsIgnoreCase("p"))
+            {
+                HTMLUtils.stripFirstElementInside(document, HTMLConstants.TAG_HTML, HTMLConstants.TAG_P);
+            } else {
+                throw new MacroExecutionException(
+                    "When using HTML content inline, you can only use inline HTML content."
+                        + " Block HTML content (such as tables) cannot be displayed."
+                        + " Try leaving an empty line before and after the macro.");
+            }
+        }
+
+        // Don't print the XML declaration nor the XHTML DocType.
+        cleanedContent = HTMLUtils.toString(document, true, true);
+
+        // Don't print the top level html element (which is always present and at the same location
+        // since it's been normalized by the HTML cleaner)
+        // Note: we trim the first 7 characters since they correspond to a leading new line (generated by
+        // XMLUtils.toString() since the doctype is printed on a line by itself followed by a new line) +
+        // the 6 chars from "<html>".
+        cleanedContent = cleanedContent.substring(7, cleanedContent.length() - 8);
```

Comment:
Sounds a bit dangerous on the long term: shouldn't we check that the cleanedContent does start with `<html>\n` before performing this? and maybe throwing an exception if the content is not exactly the same? I'm a bit afraid that we might break everything if for example a lib change its behaviour to not use a new line anymore.

Reply:
Yes, I also don't like this code, I've just copied it from the HTML macro.

Reply:
Note that I actually changed it a bit in https://github.com/xwiki/xwiki-rendering/pull/219/files#diff-d03e09b8808925132b8ef48ce07ef681ef58c4cb5f28cf00bd0b9284051c2d0dR149.